### ISSUE TITLE
Double not incorrectly triggers `nonminimal_bool` when !v returns `bool` but v is not `bool`

### ISSUE THREAD

**Issue by endorpersand** [rust-lang/rust-clippy#10836]
### Summary

When a type implements `Not` and outputs `bool`, Clippy suggests to simplify the expression `!!x` to just `x`, even though that isn't equivalent.

### Lint Name

nonminimal_bool

### Reproducer

I tried this code:

```rust
struct Foo(bool);
impl std::ops::Not for Foo {
    type Output = bool;

    fn not(self) -> Self::Output {
        !self.0
    }
}

fn main() {
    let _: bool = !!Foo(true);
}
```

I saw this happen:

```
warning: this boolean expression can be simplified
  --> src/report.rs:11:19
   |
11 |     let _: bool = !!Foo(true);
   |                   ^^^^^^^^^^^ help: try: `Foo(true)`
   |
```

I expected to see this happen:
This lint shouldn't have triggered. 

### Version

```text
rustc 1.69.0 (84c898d65 2023-04-16)
binary: rustc
commit-hash: 84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc
commit-date: 2023-04-16
host: x86_64-apple-darwin
release: 1.69.0
LLVM version: 15.0.7
```

**Comment by xFrednet** [comment]
This probably only requires a type check for the inner expression and some tests. It should be a good first issue :)

@rustbot label +good-first-issue

**Comment by lochetti** [comment]
Maybe we can close this issue, after the PR https://github.com/rust-lang/rust-clippy/pull/10845?
